### PR TITLE
Light History Store

### PR DESCRIPTION
**Pull request by viquezclaudio**
Implementation and refactoring of the Light/Regular HistoryStore

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: blockchain/src/history/light_history_store.rs [hunk 1]
```diff
@@ -29,22 +32,65 @@ impl LightHistoryStore {
     /// Creates a new LightHistoryStore.
     pub fn _new(db: DatabaseProxy) -> Self {
```
Comment (styppo) [inline blockchain/src/history/light_history_store.rs:33]:
```suggestion
    pub fn new(db: DatabaseProxy) -> Self {
```

File: blockchain/src/history/history_store.rs [hunk 2]
```diff
@@ -437,6 +438,21 @@ impl HistoryStore {
             Some(v) => v.index,
         }
     }
+
+    /// Calculates the history tree root from a vector of historic transactions. It doesn't use the
+    /// database, it is just used to check the correctness of the history root when syncing.
+    fn _root_from_hist_txs(hist_txs: &[HistoricTransaction]) -> Option<Blake2bHash> {
```
Comment (styppo) [inline blockchain/src/history/history_store.rs:444]:
```suggestion
    fn root_from_hist_txs(hist_txs: &[HistoricTransaction]) -> Option<Blake2bHash> {
```

File: blockchain/src/history/validity_store.rs [hunk 3]
```diff
@@ -1,4 +1,55 @@
...
+use nimiq_hash::Blake2bHash;
+use nimiq_transaction::historic_transaction::RawTransactionHash;
+
 /// The validity store is used by full nodes to keep track of which
 /// transactions have occurred within the validity window without
 /// having to store the full transactions
-pub struct ValidityStore {}
+pub struct ValidityStore {
+    /// The set of all raw transactions of the current validity window.
+    pub(crate) _txs: HashSet<Blake2bHash>,
```
Comment (styppo) [inline blockchain/src/history/validity_store.rs:12]:
```suggestion
    pub(crate) txs: HashSet<Blake2bHash>,
```

File: blockchain/src/history/validity_store.rs [hunk 4]
```diff
@@ -1,4 +1,55 @@
...
+
+    pub(crate) _block_txs: HashMap<u32, HashSet<Blake2bHash>>,
```
Comment (styppo) [inline blockchain/src/history/validity_store.rs:14]:
```suggestion
    pub(crate) block_txs: HashMap<u32, HashSet<Blake2bHash>>,
```

File: blockchain/src/history/validity_store.rs [hunk 5]
```diff
@@ -1,4 +1,55 @@
...
+
+                    self._block_txs.insert(block.block_number(), block_txns);
+                }
+
+                self._txs.insert(raw_tx_hash.into());
+            }
+        }
+    }
+
+    pub(crate) fn _delete_block_transactions(&mut self, block_number: u32) {
```
Comment (styppo) [inline blockchain/src/history/validity_store.rs:48]:
```suggestion
    pub(crate) fn delete_block_transactions(&mut self, block_number: u32) {
```

File: blockchain/src/history/validity_store.rs [hunk 6]
```diff
@@ -1,4 +1,55 @@
...
+
+impl ValidityStore {
+    pub(crate) fn _new() -> Self {
+        Self {
+            _txs: HashSet::new(),
+            _block_txs: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn _has_transaction(&self, raw_tx_hash: RawTransactionHash) -> bool {
```
Comment (styppo) [inline blockchain/src/history/validity_store.rs:25]:
```suggestion
    pub(crate) fn has_transaction(&self, raw_tx_hash: RawTransactionHash) -> bool {
```

File: blockchain/src/history/validity_store.rs [hunk 7]
```diff
@@ -1,4 +1,55 @@
...
+        self._txs.contains(&raw_tx_hash.into())
+    }
+
+    pub(crate) fn _add_block_transactions(&mut self, block: &Block) {
```
Comment (styppo) [inline blockchain/src/history/validity_store.rs:29]:
```suggestion
    pub(crate) fn add_block_transactions(&mut self, block: &Block) {
```

**Review by styppo — COMMENTED** [review]
LGTM. Just wondering about all those `_` prefixes for stuff that is not unused.

File: blockchain/src/history/history_store.rs (hunk 2 above)
Comment (jsdanielh) [reply blockchain/src/history/history_store.rs:444]: Yeah, why do we need to underscore it?

File: blockchain/src/history/light_history_store.rs (hunk 1 above)
Comment (jsdanielh) [reply blockchain/src/history/light_history_store.rs:33]: Yeah, why do we need to underscore it?

File: blockchain/src/history/validity_store.rs (hunk 4 above)
Comment (jsdanielh) [reply blockchain/src/history/validity_store.rs:14]: Missing rustdoc and why is it underscored?

**Comment by viquezclaudio** [comment]
Functions are underscored because they will be used in a subsequent PR.
I would like to merge this PR first, without introducing warnings (because of the unused functions) and then create the next PR

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**